CountingSort/app_resources/prefix_sum_shader.comp.hlsl

```diff
@@ -60,22 +60,28 @@ void main(uint32_t3 ID : SV_GroupThreadID, uint32_t3 GroupID : SV_GroupID)
     nbl::hlsl::glsl::barrier();
 
     uint32_t sum = 0;
+    uint32_t scan_sum = 0;
     
     for (int i = 0; i < BucketsPerThread; i++)
     {
-        sum = nbl::hlsl::workgroup::inclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
+        sum = nbl::hlsl::workgroup::exclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
         template __call <ScratchProxy>
         (sdata[WorkgroupSize * i + tid], arithmeticAccessor);
-        
+
         arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
     
         nbl::hlsl::glsl::atomicAdd(scratch[WorkgroupSize * i + tid], sum);
+        if ((tid == WorkgroupSize - 1) && i > 0)
+            nbl::hlsl::glsl::atomicAdd(scratch[WorkgroupSize * i], scan_sum);
         
         arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
         
         if ((tid == WorkgroupSize - 1) && i < (BucketsPerThread - 1))
-            sdata[WorkgroupSize * (i + 1)] += sum;
-    
+        {
+            scan_sum = sum + sdata[WorkgroupSize * i + tid];
+            sdata[WorkgroupSize * (i + 1)] += scan_sum;
+        }
+
```

Comment:
this is quite convoluted, you'd need to revert to `inclusive_scan` though, then you can do this
```cpp
uint32_t lastResult;
{
   uint32_t scanInput = scratch[WorkgroupSize * i + tid];
   if (i>0) // a value added to the first entry of the scan will propagate throughout
      scanInput = lastResult;
   acccessor.barrier();
   uint32_t result = inclusive_scan(scanInput,accessor);
   nbl::hlsl::glsl::atomicAdd(scratch[WorkgroupSize * i + tid], result);
   lastResult = result;
}
```

if you want an exclusive scan in the scatter pass, you can just read the `scratch` at offset +1

Reply:
Aren't we completely ignoring the sdata[] array that contains the data that needs to be summed & uploaded to scratch? And for i>0, it seems to ignore the values of the current iteration for the previous result.

Reply:
lets move the conversation to the nabla PR, yes I've made a mistake here cause last invocation needs to send the sym to the first, my new pseudocode accounts for that